File: Java/Block.java
import java.security.MessageDigest;
import java.util.Date;

public class Block {
    public int index;
    public String timestamp;
    public String data;
    public String previousHash;
    public int nonce;
    public String hash;

    public Block(int index, String data, String previousHash) {
        this.index = index;
        this.timestamp = new Date().toString();
        this.data = data;
        this.previousHash = previousHash;
        this.nonce = 0;
        this.hash = calculateHash();
    }

    public String calculateHash() {
        try {
            String input = index + timestamp + data + previousHash + nonce;
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashBytes = digest.digest(input.getBytes("UTF-8"));
            StringBuilder hex = new StringBuilder();
            for (byte b : hashBytes)
                hex.append(String.format("%02x", b));
            return hex.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public void mineBlock(int difficulty) {
        String target = "0".repeat(difficulty);
        while (!hash.substring(0, difficulty).equals(target)) {
            nonce++;
            hash = calculateHash();
        }
        System.out.println("Block " + index + " mined: " + hash);
    }
}

File: block.py
# block.py
import hashlib, time

class Block:
    def __init__(self, index, data, prev_hash='0', difficulty=0):
        self.index = index
        self.data = data
        self.timestamp = time.time()  # Use float timestamp
        self.prev_hash = prev_hash
        self.nonce = 0
        self.hash = self.calc_hash()

        if difficulty > 0:
            self.mine(difficulty)

    def calc_hash(self):
        text = f"{self.index}{self.timestamp}{self.data}{self.prev_hash}{self.nonce}"
        return hashlib.sha256(text.encode()).hexdigest()

    def mine(self, diff):
        while not self.hash.startswith('0' * diff):
            self.nonce += 1
            self.hash = self.calc_hash()
        print(f"Block {self.index} mined: {self.hash}")